File: node/src/providers/request-plan.ts
import { createHash } from "node:crypto"
import { createRequire } from "node:module"
import type { RenderedContext, ToolSchema, ProviderUsage } from "../types.js"
import type { GenerationProtocol } from "./protocol-capabilities.js"

export interface ProviderRequestEndpoint {
  id: string
  protocol: string
  baseURL: string
}

/** One provider-visible request, deliberately excluding credentials and transport-only retries. */
export interface ProviderRequestPlan {
  providerId: string
  modelId: string
  endpoint: ProviderRequestEndpoint
  context: RenderedContext
  tools: ToolSchema[]
  options: Record<string, unknown>
  fingerprint: string
  stablePrefixFingerprint: string
}

export interface NormalizedProviderUsage extends ProviderUsage {
  /** Input not accounted as a cache read or write. `inputTokens` remains the full footprint. */
  uncachedInputTokens: number
}

/** Durable host fact: a specific logical request was counted before execution. */
export interface RecordedPromptMeasurement {
  requestFingerprint: string
  inputTokens: number
  source:
    | { kind: "native"; provider: string }
    | { kind: "local_exact"; tokenizer: string }
    | { kind: "postflight" }
    | { kind: "heuristic" }
  confidence: "exact" | "high_confidence" | "low_confidence"
}

export interface PricingSnapshot {
  version: string
  currency: string
  region: string
  effectiveFrom: string
  expiresAt?: string
  ratesPerMillion: {
    input: number
    output: number
    cacheRead?: number
    cacheCreation?: number
    reasoning?: number
  }
}

export type CostObservation =
  | { source: "snapshot"; currency: string; amount: number; pricingVersion: string }
  | { source: "unpriced"; reason: "pricing_snapshot_not_effective" | "pricing_snapshot_expired" | "invalid_pricing_snapshot" }

const TRANSPORT_ONLY_KEYS = new Set([
  "apiKey", "api_key", "bearerToken", "bearer_token", "authorization", "credential",
  "credentials", "retry", "maxRetries", "baseDelay", "timeout", "signal", "access_token", "refresh_token", "token", "secret", "x-api-key",
])

export function createProviderRequestPlan(input: Omit<ProviderRequestPlan, "fingerprint" | "stablePrefixFingerprint" | "options"> & {
  options?: Record<string, unknown>
}): ProviderRequestPlan {
  const options = materialOptions(input.options ?? {})
  const plan = {
    providerId: input.providerId,
    modelId: input.modelId,
    endpoint: sanitizeEndpoint(input.endpoint),
    context: clone(input.context),
    tools: clone(input.tools),
    options,
  }
  const stablePrefix = {
    providerId: plan.providerId,
    modelId: plan.modelId,
    endpoint: plan.endpoint,
    context: stablePrefixContext(plan.context),
    tools: plan.tools,
    options: plan.options,
  }
  return {
    ...plan,
    fingerprint: sha256(stableJson(plan)),
    stablePrefixFingerprint: sha256(stableJson(stablePrefix)),
  }
}

/** Build the plan from a resolved provider when the runner only has the public provider object. */
export function createProviderRequestPlanForProvider(
  provider: {
    descriptor?(): { provider: string; protocol: string; model: string }
    requestPlanIdentity?(): {
      providerId?: string
      modelId?: string
      endpoint?: { id?: string; protocol?: string; baseURL?: string }
    }
  },
  context: RenderedContext,
  tools: ToolSchema[],
  options?: Record<string, unknown>,
): ProviderRequestPlan {
  const descriptor = provider.descriptor?.() ?? { provider: "unknown", protocol: "unknown", model: "unknown" }
  const identity = provider.requestPlanIdentity?.()
  return createProviderRequestPlan({
    providerId: identity?.providerId ?? descriptor.provider,
    modelId: identity?.modelId ?? descriptor.model,
    endpoint: {
      id: identity?.endpoint?.id ?? `${descriptor.provider}.${descriptor.protocol}`,
      protocol: identity?.endpoint?.protocol ?? descriptor.protocol,
      baseURL: identity?.endpoint?.baseURL ?? "",
    },
    context,
    tools,
    options,
  })
}

/**
 * P4 §1.3: the resolved route one provider execution is pinned to. Content-addressed
 * (`routeId` = stable hash of every field except `capabilitiesRef`); the capability table's
 * authority stays the static vocabulary in protocol-capabilities.ts — this holds a reference,
 * never a copy.
 */
export interface ResolvedProviderRoute {
  routeId: string
  provider: string
  protocol: GenerationProtocol
  model: string
  endpoint: ProviderRequestEndpoint
  /** Adapter implementation version (the @deepstrike/sdk package version; F8). */
  adapterVersion: string
  /** Capability snapshot reference: the protocol constant name (e.g. "anthropic-messages").
   *  A provider exposing capability overrides would append a summary here — none does today. */
  capabilitiesRef: string
}

const KNOWN_GENERATION_PROTOCOLS = new Set<GenerationProtocol>([
  "anthropic-messages",
  "openai-chat",
  "openai-responses",
  "gemini",
  "ollama-chat",
])

/**
 * P4-S1: assemble the run's route once at runner construction (P4 §0.2 — the provider is
 * fixed for the run today; a future failover resolver re-evaluates per attempt with the same
 * shape). Never throws: evidence assembly degrades to "unknown" fields rather than breaking a run.
 */
export function resolveProviderRoute(provider: {
  descriptor?(): { provider: string; protocol: string; model: string }
  requestPlanIdentity?(): {
    providerId?: string
    modelId?: string
    endpoint?: { id?: string; protocol?: string; baseURL?: string }
  }
}): ResolvedProviderRoute {
  try {
    const descriptor = provider.descriptor?.() ?? { provider: "unknown", protocol: "unknown", model: "unknown" }
    const identity = provider.requestPlanIdentity?.()
    const protocolRaw = identity?.endpoint?.protocol ?? descriptor.protocol
    // Evidence records what the descriptor said, even when a foreign provider speaks a protocol
    // outside the in-tree vocabulary — the field is a report, not a gate.
    const protocol = protocolRaw as GenerationProtocol
    const route = {
      provider: identity?.providerId ?? descriptor.provider,
      protocol,
      model: identity?.modelId ?? descriptor.model,
      endpoint: sanitizeEndpoint({
        id: identity?.endpoint?.id ?? `${descriptor.provider}.${descriptor.protocol}`,
        protocol: protocolRaw,
        baseURL: identity?.endpoint?.baseURL ?? "",
      }),
      adapterVersion: adapterPackageVersion(),
      capabilitiesRef: KNOWN_GENERATION_PROTOCOLS.has(protocol) ? protocol : "unknown",
    }
    // §1.3: content addressing covers every field EXCEPT capabilitiesRef (a reference, not content).
    const { capabilitiesRef: _ref, ...addressed } = route
    return { ...route, routeId: sha256(stableJson(addressed)) }
  } catch {
    const fallback = {
      provider: "unknown",
      protocol: "unknown" as GenerationProtocol,
      model: "unknown",
      endpoint: { id: "unknown", protocol: "unknown", baseURL: "" },
      adapterVersion: "unknown",
    }
    return { ...fallback, capabilitiesRef: "unknown", routeId: sha256(stableJson(fallback)) }
  }
}

/** The adapter's package version, read from package.json at runtime. "unknown" when unreadable —
 *  route assembly is evidence plumbing and must never throw (B7: evidence, not kernel input). */
function adapterPackageVersion(): string {
  try {
    // From both src/providers/ (ts) and dist/providers/ (js), ../../package.json is the SDK manifest.
    const require = createRequire(import.meta.url)
    const manifest = require("../../package.json") as { version?: unknown }
    return typeof manifest.version === "string" && manifest.version.length > 0 ? manifest.version : "unknown"
  } catch {
    return "unknown"
  }
}

export function estimateProviderPromptTokens(context: RenderedContext, tools: ToolSchema[]): number {
  const bytes = new TextEncoder().encode(stableJson({ context, tools })).byteLength
  return Math.max(1, Math.ceil(bytes / 4))
}

/** Bind a preflight count to its exact provider-visible request. Replay only reuses matching facts. */
export function recordPromptMeasurement(
  plan: Pick<ProviderRequestPlan, "fingerprint">,
  measurement: Omit<RecordedPromptMeasurement, "requestFingerprint">,
): RecordedPromptMeasurement {
  return {
    requestFingerprint: plan.fingerprint,
    inputTokens: requireNonNegativeInteger(measurement.inputTokens, "inputTokens"),
    source: clone(measurement.source),
    confidence: measurement.confidence,
  }
}

export function measurementForPlan(
  plan: Pick<ProviderRequestPlan, "fingerprint">,
  recorded: RecordedPromptMeasurement | undefined,
): RecordedPromptMeasurement | undefined {
  if (!recorded || recorded.requestFingerprint !== plan.fingerprint) return undefined
  if (!Number.isSafeInteger(recorded.inputTokens) || recorded.inputTokens < 0) return undefined
  if (recorded.confidence !== "exact" && recorded.confidence !== "high_confidence" && recorded.confidence !== "low_confidence") return undefined
  const source = recorded.source
  if (!source || typeof source !== "object") return undefined
  if (source.kind === "native" && typeof source["provider"] === "string" && source["provider"].length > 0) return clone(recorded)
  if (source.kind === "local_exact" && typeof source.tokenizer === "string" && source.tokenizer.length > 0) return clone(recorded)
  if (source.kind === "postflight") return clone(recorded)
  if (source.kind === "heuristic") return clone(recorded)
  return undefined
}

/** Normalize postflight provider facts without turning estimates into actual usage. */
export function normalizeProviderUsage(usage: ProviderUsage): NormalizedProviderUsage {
  const inputTokens = requireNonNegativeInteger(usage.inputTokens, "inputTokens")
  const outputTokens = requireNonNegativeInteger(usage.outputTokens, "outputTokens")
  const cacheReadInputTokens = optionalNonNegativeInteger(usage.cacheReadInputTokens, "cacheReadInputTokens")
  const cacheCreationInputTokens = optionalNonNegativeInteger(usage.cacheCreationInputTokens, "cacheCreationInputTokens")
  const reasoningTokens = optionalNonNegativeInteger(usage.reasoningTokens, "reasoningTokens")
  const cached = (cacheReadInputTokens ?? 0) + (cacheCreationInputTokens ?? 0)
  if (cached > inputTokens) throw new RangeError("cache token subsets cannot exceed inputTokens")
  if (reasoningTokens !== undefined && reasoningTokens > outputTokens) {
    throw new RangeError("reasoningTokens must be a subset of outputTokens")
  }
  return {
    inputTokens,
    uncachedInputTokens: inputTokens - cached,
    outputTokens,
    ...(cacheReadInputTokens !== undefined ? { cacheReadInputTokens } : {}),
    ...(cacheCreationInputTokens !== undefined ? { cacheCreationInputTokens } : {}),
    ...(reasoningTokens !== undefined ? { reasoningTokens } : {}),
  }
}

/** Cost is derived only from an explicit, time-valid host snapshot; otherwise it stays unknown. */
export function priceProviderUsage(
  usage: NormalizedProviderUsage,
  snapshot: PricingSnapshot,
  observedAt: string | Date = new Date(),
): CostObservation {
  const at = typeof observedAt === "string" ? new Date(observedAt) : observedAt
  const from = new Date(snapshot.effectiveFrom)
  const expires = snapshot.expiresAt ? new Date(snapshot.expiresAt) : undefined
  const rates = snapshot.ratesPerMillion
  const requiredRates = [rates.input, rates.output]
  if (!snapshot.version || !snapshot.currency || Number.isNaN(at.valueOf()) || Number.isNaN(from.valueOf())
    || requiredRates.some(rate => typeof rate !== "number" || !Number.isFinite(rate) || rate < 0)
    || Object.values(rates).some(rate => typeof rate !== "number" || !Number.isFinite(rate) || rate < 0)) {
    return { source: "unpriced", reason: "invalid_pricing_snapshot" }
  }
  if (at < from) return { source: "unpriced", reason: "pricing_snapshot_not_effective" }
  if (expires && (Number.isNaN(expires.valueOf()) || at >= expires)) {
    return { source: "unpriced", reason: "pricing_snapshot_expired" }
  }
  const amount = (
    usage.uncachedInputTokens * rates.input
    + usage.outputTokens * rates.output
    + (usage.cacheReadInputTokens ?? 0) * (rates.cacheRead ?? rates.input)
    + (usage.cacheCreationInputTokens ?? 0) * (rates.cacheCreation ?? rates.input)
    + (usage.reasoningTokens ?? 0) * (rates.reasoning ?? 0)
  ) / 1_000_000
  return { source: "snapshot", currency: snapshot.currency, amount, pricingVersion: snapshot.version }
}

function materialOptions(options: Record<string, unknown>): Record<string, unknown> {
  return sanitizeMaterialValue(options) as Record<string, unknown>
}

function sanitizeEndpoint(endpoint: ProviderRequestEndpoint): ProviderRequestEndpoint {
  try {
    const url = new URL(endpoint.baseURL)
    url.username = ""
    url.password = ""
    url.search = ""
    url.hash = ""
    return { ...clone(endpoint), baseURL: url.toString().replace(/\/$/, "") }
  } catch {
    return { ...clone(endpoint), baseURL: "" }
  }
}

function sanitizeMaterialValue(value: unknown): unknown {
  if (value === undefined || typeof value === "function") return undefined
  if (Array.isArray(value)) return value.map(sanitizeMaterialValue).filter(item => item !== undefined)
  if (value !== null && typeof value === "object") {
    return Object.fromEntries(Object.keys(value as Record<string, unknown>).sort().flatMap(key => {
      if (TRANSPORT_ONLY_KEYS.has(key) || isTransportOnlyKey(key)) return []
      const sanitized = sanitizeMaterialValue((value as Record<string, unknown>)[key])
      return sanitized === undefined ? [] : [[key, sanitized]]
    }))
  }
  return value
}

function isTransportOnlyKey(key: string): boolean {
  const normalized = key.toLowerCase().replace(/[^a-z0-9]/g, "")
  return normalized.includes("authorization") || normalized.includes("credential")
    || normalized.includes("accesstoken") || normalized.includes("refreshtoken")
    || normalized.includes("apikey") || normalized === "bearer" || normalized === "token"
    || normalized === "secret" || normalized === "xapikey"
}

function stableJson(value: unknown): string {
  if (value === null || typeof value !== "object") return JSON.stringify(value)
  if (Array.isArray(value)) return `[${value.map(stableJson).join(",")}]`
  const object = value as Record<string, unknown>
  return `{${Object.keys(object).sort().map(key => `${JSON.stringify(key)}:${stableJson(object[key])}`).join(",")}}`
}

function stablePrefixContext(context: RenderedContext): Record<string, unknown> {
  const frozenPrefixLen = context.frozenPrefixLen ?? 0
  return {
    systemText: context.systemText,
    ...(context.systemStable !== undefined ? { systemStable: context.systemStable } : {}),
    ...(context.systemKnowledge !== undefined ? { systemKnowledge: context.systemKnowledge } : {}),
    frozenPrefixLen,
    turns: clone(context.turns.slice(0, frozenPrefixLen)),
  }
}

function sha256(value: string): string {
  return `sha256:${createHash("sha256").update(value).digest("hex")}`
}

function clone<T>(value: T): T {
  if (value === undefined || value === null || typeof value !== "object") return value
  if (Array.isArray(value)) return value.map(clone) as T
  return Object.fromEntries(Object.entries(value as Record<string, unknown>).map(([key, item]) => [key, clone(item)])) as T
}

function requireNonNegativeInteger(value: number, name: string): number {
  if (!Number.isSafeInteger(value) || value < 0) throw new RangeError(`${name} must be a non-negative safe integer`)
  return value
}

function optionalNonNegativeInteger(value: number | undefined, name: string): number | undefined {
  return value === undefined ? undefined : requireNonNegativeInteger(value, name)
}

File: rust/src/providers/request_plan.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;
use sha2::{Digest, Sha256};
use time::OffsetDateTime;

use deepstrike_core::context::measurement::{MeasurementConfidence, MeasurementSource};

/// Provider-visible endpoint identity. Credentials and retry transport state are intentionally
/// excluded from this host-side request contract.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct ProviderRequestEndpoint {
    pub id: String,
    pub protocol: String,
    #[serde(rename = "baseURL")]
    pub base_url: String,
}

impl ProviderRequestEndpoint {
    pub fn new(
        id: impl Into<String>,
        protocol: impl Into<String>,
        base_url: impl Into<String>,
    ) -> Self {
        Self {
            id: id.into(),
            protocol: protocol.into(),
            base_url: base_url.into(),
        }
    }
}

/// One material provider request. The fingerprint binds a preflight measurement to exactly the
/// model, endpoint, context, tools, and material request options that will be executed.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct ProviderRequestPlan {
    pub version: u8,
    pub provider_id: String,
    pub model_id: String,
    pub endpoint: ProviderRequestEndpoint,
    pub context: Value,
    pub tools: Vec<Value>,
    pub options: Value,
    pub fingerprint: String,
}

impl ProviderRequestPlan {
    pub fn new(
        provider_id: impl Into<String>,
        model_id: impl Into<String>,
        endpoint: ProviderRequestEndpoint,
        context: Value,
        tools: Vec<Value>,
        options: Value,
    ) -> Result<Self, RequestPlanError> {
        let provider_id = provider_id.into();
        let model_id = model_id.into();
        let options = material_options(options)?;
        let hashed = serde_json::json!({
            "version": 1,
            "providerId": provider_id,
            "modelId": model_id,
            "endpoint": {
                "id": endpoint.id,
                "protocol": endpoint.protocol,
                "baseURL": endpoint.base_url,
            },
            "context": context,
            "tools": tools,
            "options": options,
        });
        let fingerprint = format!(
            "sha256:{:x}",
            Sha256::digest(canonical_json(&hashed).as_bytes())
        );
        Ok(Self {
            version: 1,
            provider_id,
            model_id,
            endpoint,
            context: hashed["context"].clone(),
            tools: hashed["tools"].as_array().cloned().unwrap_or_default(),
            options: hashed["options"].clone(),
            fingerprint,
        })
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct RecordedPromptMeasurement {
    pub version: u8,
    pub request_fingerprint: String,
    pub input_tokens: u64,
    pub source: MeasurementSource,
    pub confidence: MeasurementConfidence,
}

pub type PromptMeasurementSource = MeasurementSource;

pub fn record_prompt_measurement(
    plan: &ProviderRequestPlan,
    input_tokens: u64,
    source: MeasurementSource,
    confidence: MeasurementConfidence,
) -> RecordedPromptMeasurement {
    RecordedPromptMeasurement {
        version: 1,
        request_fingerprint: plan.fingerprint.clone(),
        input_tokens,
        source,
        confidence,
    }
}

pub fn measurement_for_plan(
    plan: &ProviderRequestPlan,
    measurement: &RecordedPromptMeasurement,
) -> Option<RecordedPromptMeasurement> {
    (measurement.version == 1 && measurement.request_fingerprint == plan.fingerprint)
        .then(|| measurement.clone())
}

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct ProviderUsage {
    pub input_tokens: u64,
    pub output_tokens: u64,
    pub cache_read_input_tokens: u64,
    pub cache_creation_input_tokens: u64,
    pub reasoning_tokens: Option<u64>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct NormalizedProviderUsage {
    pub input_tokens: u64,
    pub uncached_input_tokens: u64,
    pub output_tokens: u64,
    pub cache_read_input_tokens: u64,
    pub cache_creation_input_tokens: u64,
    pub reasoning_tokens: Option<u64>,
}

pub fn normalize_provider_usage(
    usage: &ProviderUsage,
) -> Result<NormalizedProviderUsage, RequestPlanError> {
    let cached = usage
        .cache_read_input_tokens
        .checked_add(usage.cache_creation_input_tokens)
        .ok_or(RequestPlanError::InvalidUsage)?;
    if cached > usage.input_tokens
        || usage
            .reasoning_tokens
            .is_some_and(|tokens| tokens > usage.output_tokens)
    {
        return Err(RequestPlanError::InvalidUsage);
    }
    Ok(NormalizedProviderUsage {
        input_tokens: usage.input_tokens,
        uncached_input_tokens: usage.input_tokens - cached,
        output_tokens: usage.output_tokens,
        cache_read_input_tokens: usage.cache_read_input_tokens,
        cache_creation_input_tokens: usage.cache_creation_input_tokens,
        reasoning_tokens: usage.reasoning_tokens,
    })
}

#[derive(Debug, Clone, PartialEq)]
pub struct PricingRates {
    pub input: f64,
    pub output: f64,
    pub cache_read: Option<f64>,
    pub cache_creation: Option<f64>,
    pub reasoning: Option<f64>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct PricingSnapshot {
    pub version: String,
    pub currency: String,
    pub region: String,
    pub effective_from: String,
    pub expires_at: Option<String>,
    pub rates_per_million: PricingRates,
}

impl PricingSnapshot {
    pub fn new(
        version: impl Into<String>,
        currency: impl Into<String>,
        region: impl Into<String>,
        effective_from: impl Into<String>,
        expires_at: Option<String>,
        rates_per_million: PricingRates,
    ) -> Self {
        Self {
            version: version.into(),
            currency: currency.into(),
            region: region.into(),
            effective_from: effective_from.into(),
            expires_at,
            rates_per_million,
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum CostObservation {
    Snapshot {
        currency: String,
        amount: f64,
        pricing_version: String,
    },
    Unpriced {
        reason: UnpricedReason,
    },
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum UnpricedReason {
    SnapshotNotEffective,
    SnapshotExpired,
    InvalidPricingSnapshot,
}

pub fn price_provider_usage(
    usage: &NormalizedProviderUsage,
    snapshot: &PricingSnapshot,
    observed_at: &str,
) -> CostObservation {
    let valid_rates = [
        Some(snapshot.rates_per_million.input),
        Some(snapshot.rates_per_million.output),
        snapshot.rates_per_million.cache_read,
        snapshot.rates_per_million.cache_creation,
        snapshot.rates_per_million.reasoning,
    ]
    .into_iter()
    .flatten()
    .all(|rate| rate.is_finite() && rate >= 0.0);
    let observed_at = parse_rfc3339(observed_at);
    let effective_from = parse_rfc3339(&snapshot.effective_from);
    let expires_at = snapshot
        .expires_at
        .as_deref()
        .map(parse_rfc3339)
        .transpose();
    if snapshot.version.is_empty()
        || snapshot.currency.is_empty()
        || !valid_rates
        || observed_at.is_err()
        || effective_from.is_err()
        || expires_at.is_err()
    {
        return CostObservation::Unpriced {
            reason: UnpricedReason::InvalidPricingSnapshot,
        };
    }
    let observed_at = observed_at.expect("checked above");
    let effective_from = effective_from.expect("checked above");
    if observed_at < effective_from {
        return CostObservation::Unpriced {
            reason: UnpricedReason::SnapshotNotEffective,
        };
    }
    if expires_at
        .expect("checked above")
        .is_some_and(|expires| observed_at >= expires)
    {
        return CostObservation::Unpriced {
            reason: UnpricedReason::SnapshotExpired,
        };
    }
    let rates = &snapshot.rates_per_million;
    let amount = (usage.uncached_input_tokens as f64 * rates.input
        + usage.output_tokens as f64 * rates.output
        + usage.cache_read_input_tokens as f64 * rates.cache_read.unwrap_or(rates.input)
        + usage.cache_creation_input_tokens as f64 * rates.cache_creation.unwrap_or(rates.input)
        + usage.reasoning_tokens.unwrap_or_default() as f64 * rates.reasoning.unwrap_or(0.0))
        / 1_000_000.0;
    CostObservation::Snapshot {
        currency: snapshot.currency.clone(),
        amount,
        pricing_version: snapshot.version.clone(),
    }
}

fn parse_rfc3339(value: &str) -> Result<OffsetDateTime, time::error::Parse> {
    OffsetDateTime::parse(value, &time::format_description::well_known::Rfc3339)
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, thiserror::Error)]
pub enum RequestPlanError {
    #[error("request plan options must be a JSON object")]
    OptionsMustBeObject,
    #[error("provider usage has inconsistent token subsets")]
    InvalidUsage,
}

fn material_options(options: Value) -> Result<Value, RequestPlanError> {
    let Value::Object(object) = options else {
        return Err(RequestPlanError::OptionsMustBeObject);
    };
    Ok(sanitize_value(Value::Object(object)).unwrap_or(Value::Object(Default::default())))
}

fn sanitize_value(value: Value) -> Option<Value> {
    match value {
        Value::Array(values) => Some(Value::Array(
            values.into_iter().filter_map(sanitize_value).collect(),
        )),
        Value::Object(values) => Some(Value::Object(
            values
                .into_iter()
                .filter_map(|(key, value)| {
                    (!transport_only_key(&key))
                        .then(|| sanitize_value(value).map(|value| (key, value)))
                        .flatten()
                })
                .collect(),
        )),
        value => Some(value),
    }
}

fn transport_only_key(key: &str) -> bool {
    let normalized: String = key
        .chars()
        .filter(|character| character.is_ascii_alphanumeric())
        .flat_map(|character| character.to_lowercase())
        .collect();
    matches!(normalized.as_str(), "retry" | "maxretries" | "basedelay" | "timeout" | "signal")
        || normalized.contains("authorization")
        || normalized.contains("credential")
        || normalized.contains("accesstoken")
        || normalized.contains("refreshtoken")
        || normalized.contains("apikey")
        || matches!(normalized.as_str(), "bearer" | "token" | "secret" | "xapikey")
}

fn canonical_json(value: &Value) -> String {
    match value {
        Value::Null => "null".into(),
        Value::Bool(value) => value.to_string(),
        Value::Number(value) => value.to_string(),
        Value::String(value) => serde_json::to_string(value).expect("strings serialize"),
        Value::Array(values) => format!(
            "[{}]",
            values
                .iter()
                .map(canonical_json)
                .collect::<Vec<_>>()
                .join(",")
        ),
        Value::Object(values) => format!(
            "{{{}}}",
            values
                .iter()
                .map(|(key, value)| format!(
                    "{}:{}",
                    serde_json::to_string(key).expect("keys serialize"),
                    canonical_json(value)
                ))
                .collect::<Vec<_>>()
                .join(","),
        ),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn uses_the_shared_cross_sdk_sha256_fixture() {
        let fixture: serde_json::Value = serde_json::from_str(include_str!(
            "../../../tests/fixtures/provider-request-plan/v1.json"
        ))
        .expect("fixture is valid JSON");
        let input = fixture.get("input").expect("fixture input");
        let endpoint = input.get("endpoint").expect("fixture endpoint");
        let plan = ProviderRequestPlan::new(
            input["providerId"].as_str().unwrap(),
            input["modelId"].as_str().unwrap(),
            ProviderRequestEndpoint::new(
                endpoint["id"].as_str().unwrap(),
                endpoint["protocol"].as_str().unwrap(),
                endpoint["baseURL"].as_str().unwrap(),
            ),
            input["context"].clone(),
            input["tools"].as_array().unwrap().clone(),
            input["options"].clone(),
        )
        .expect("valid request plan");

        assert_eq!(plan.fingerprint, fixture["fingerprint"].as_str().unwrap());
        assert_eq!(
            plan.options,
            serde_json::json!({"auth": {"mode": "request"}, "temperature": 0.2, "transport": {}})
        );
        assert!(
            !serde_json::to_string(&plan)
                .unwrap()
                .contains("must-not-hash")
        );
    }

    #[test]
    fn records_measurements_and_prices_only_valid_snapshots() {
        let plan = ProviderRequestPlan::new(
            "openai",
            "gpt-4o",
            ProviderRequestEndpoint::new("openai.chat", "openai-chat", "https://api.openai.com/v1"),
            serde_json::json!({"systemText": "s", "turns": []}),
            vec![],
            serde_json::json!({}),
        )
        .unwrap();
        let measurement = record_prompt_measurement(
            &plan,
            42,
            PromptMeasurementSource::Native {
                provider: "openai".into(),
            },
            MeasurementConfidence::Exact,
        );
        assert_eq!(measurement_for_plan(&plan, &measurement), Some(measurement));

        let usage = normalize_provider_usage(&ProviderUsage {
            input_tokens: 120,
            output_tokens: 30,
            cache_read_input_tokens: 20,
            cache_creation_input_tokens: 10,
            reasoning_tokens: Some(6),
        })
        .unwrap();
        assert_eq!(usage.uncached_input_tokens, 90);
        assert_eq!(
            price_provider_usage(
                &usage,
                &PricingSnapshot::new(
                    "v1",
                    "USD",
                    "global",
                    "2026-08-01T00:00:00Z",
                    Some("2026-09-01T00:00:00Z".into()),
                    PricingRates {
                        input: 2.0,
                        output: 8.0,
                        cache_read: Some(0.2),
                        cache_creation: Some(2.5),
                        reasoning: None,
                    },
                ),
                "2026-08-13T00:00:00Z",
            ),
            CostObservation::Snapshot {
                currency: "USD".into(),
                amount: 0.000449,
                pricing_version: "v1".into()
            },
        );
    }

    #[test]
    fn redacts_nested_case_and_separator_variant_credentials() {
        let plan = ProviderRequestPlan::new(
            "openai",
            "gpt-4o",
            ProviderRequestEndpoint::new("openai.chat", "openai-chat", "https://api.openai.com/v1"),
            serde_json::json!({"turns": []}),
            vec![],
            serde_json::json!({"headers": {"Authorization": "secret", "x-api-key": "secret"}, "access_token": "secret", "temperature": 0.2}),
        )
        .unwrap();
        assert_eq!(plan.options, serde_json::json!({"temperature": 0.2, "headers": {}}));
        assert!(!serde_json::to_string(&plan).unwrap().contains("secret"));
    }
}
